tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_runner.cc

```diff
@@ -0,0 +1,299 @@
+  scale_offset_desc.set_layout(se::dnn::DataLayout::kBatchDepthYX)
+      .set_feature_map_count(input_desc.feature_map_count())
+      .set_height(1)
+      .set_width(1)
+      .set_count(1);
+
+  return std::make_pair(input_desc, scale_offset_desc);
+}
+
+void AssignCommonParams(const HloInstruction* batchnorm,
+                        CudnnBatchNormParamsCommon* params,
+                        const se::DeviceMemoryBase& operand,
+                        const se::DeviceMemory<float>& scale, float epsilon,
+                        int64 feature_index) {
+  // TF_ASSIGN_OR_RETURN(params->kind, GetCudnnBatchNormKind(batchnorm));
+  // The BatchNormTraining HLO outputs a tuple of three elements: output data,
+  // batch mean, and batch variance.  We want to make our descriptors based on
+  // the shape of the output data. Batchnorm backward call outputs a tuple of
+  // three elements: grad data, grad offset, and grad scale.  We want to make
+  // our descriptors based on the shape of the grad data.
+  const Shape& shape = batchnorm->shape().IsTuple()
+                           ? batchnorm->shape().tuple_shapes(0)
+                           : batchnorm->shape();
+  std::tie(params->operand_desc, params->scale_offset_desc) =
+      MakeBatchNormDescriptors(shape, feature_index);
+  params->operand = operand;
+  params->scale = scale;
+  params->epsilon = epsilon;
+}
+
+template <typename ElemType>
+Status RunCudnnBatchNormForwardInferenceImpl(
+    CudnnBatchNormForwardInferenceParams* params, se::Stream* stream) {
+  se::DeviceMemory<float> null_device_ptr(nullptr);
+  auto output_buf = se::DeviceMemory<ElemType>(params->output);
+  stream->ThenBatchNormalizationForward(
+      se::DeviceMemory<ElemType>(params->common.operand),
+      params->common.scale,                                         //
+      params->offset,                                               //
+      params->mean,                                                 //
+      params->variance,                                             //
+      /*side_input=*/null_device_ptr, params->common.operand_desc,  //
+      params->common.scale_offset_desc, params->common.epsilon,      //
+      se::dnn::ActivationMode::kNone,                               //
+      &output_buf,                                                  //
+      /*batch_mean=*/nullptr,                                       //
+      /*batch_var=*/nullptr,                                        //
+      /*saved_mean=*/nullptr,                                       //
+      /*saved_inv_var=*/nullptr,                                    //
+      /*is_training=*/false,                                        //
+      /*var_to_inv_var=*/nullptr,                                   //
+      /*inv_var_to_var=*/nullptr,                                   //
+      /*reserve_space_allocator=*/nullptr,                          //
+      /*workspace_allocator=*/nullptr);
+
+  return Status::OK();
```

Comment:
If we're only ever returning OK then probably the return type can be `void`?

Reply:
Correct me if I am wrong, but I believe this needs to return a Status since `RunCudnnBatchNorm<ForwardInference|ForwardTraining|Backward> returns a Status (non-OK status for the default case) back to `cudnn_batchnorm_thunk`. If I return a void here, I need to add 6 instances of `return Status::OK()`, one for each case like so:
`  PrimitiveType output_primitive_type = batchnorm->shape().element_type();
  switch (output_primitive_type) {
    case F16:
      RunCudnnBatchNormForwardInferenceImpl<Eigen::half>(
          &inference_params, stream);
      return Status::OK();
    case F32:
      RunCudnnBatchNormForwardInferenceImpl<float>(&inference_params,
                                                          stream);
      return Status::OK();
    default:
      LOG(FATAL) << batchnorm->ToString() << " is unimplemented";
  }` 
Please let me knw if I am missing something.

Reply:
Please let me knw if I am missing something.

Reply:
Or you could do:

```
  switch (output_primitive_type) {
    case F16:
      RunCudnnBatchNormForwardInferenceImpl<Eigen::half>(
          &inference_params, stream);
    case F32:
      RunCudnnBatchNormForwardInferenceImpl<float>(&inference_params,
                                                          stream);
    default:
      LOG(FATAL) << batchnorm->ToString() << " is unimplemented";
  }
  return Status::OK();
```

I don't think what you're suggesting is a good reason to have this return a `Status` (although I might be convinced if this was a heavily used function and having a trick like this saved 100s of lines).  Moreover, it makes the call site less informative; it looks like the caller could propagate an error from `RunCudnnBatchNormForwardInferenceImpl` when it always succeeds.

Reply:
Agreed!